docs/RFCS/cluster_upgrade_tool.md

```diff
@@ -0,0 +1,107 @@
+- Feature Name: Cluster Upgrade Tool
+- Status: draft
+- Start Date: 2016-09-15
+- Authors: Daniel Harrison
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: (one or more # from the issue tracker)
+
+# Summary
+
+A series of hooks to perform any necessary bookkeeping before a CockroachDB
+version upgrade.
+
+# Motivation
+
+Backward compatibility is crucial to a database, but can introduce considerable
+complexity to a codebase. All else being equal, it's better to keep backward
+compatibility logic siloed from the significant complexity that is inherent in
+software as complex as CockroachDB.
+
+# Detailed design
+
+Some versions of CockroachDB will require that a "system migration hook" be run
+before the cluster is upgraded to that version. This is expected to be
+infrequent; not every release will bump the version, only ones that need a
+migration hook.
+
+Migration hooks consist of a name, a work function, and optionally a minimum
+version. In the steady state, there is a map in code between migration names and
+the cockroach version they were released in. There is also a list of migrations
+added since the last cockroach release.
+
+When a node starts up, it checks that all known migrations have been run (the
+amount of work involved in this can be kept bounded using the `/SystemVersion`
+described below). If not, it runs them via a migration coordinator. (We should
+also do this check when a node rejoins a cluster it's been partitioned from, but
+I'm not sure how to do that.)
+
+The migration coordinator starts by heartbeating a kv entry to ensure that there
+is only ever one running at a time. Then for each missing migration, it runs the
+work function and writes a record of the completion to kv
+`/SystemVersion/<MigrationName>`. When finished, it writes its own version to
+`/SystemVersion`.
+
+The work function must be idempotent (in case a migration coordinator crashes)
+and runnable on some previous version of the cluster. If there is some minimum
+cluster version it needs, that is checked against the one written to
+`/SystemVersion`.
+
+The migration name to version map can be used to order migrations when a cluster
+needs to catch up many versions.
+
+We will introduce a thin cli wrapper `./cockroach cluster-upgrade` around
```

Comment:
It sounds like you're talking about running `cluster-upgrade` with a new binary while the cluster is still running an old version. This sounds like it could be tricky from an operational logistics perspective. It's also troubling from a security perspective since this command would presumably need an all-powerful node certificate instead of a user certificate.